lib/functions/OverloadNode.ts

```diff
@@ -0,0 +1,89 @@
+import type * as E from '../expressions';
+import type { ArgumentType } from './Core';
+
+export type SearchStack = OverloadNode[];
+
+/**
+ * Maps argument types on their specific implementation in a tree like structure.
+ */
+export class OverloadNode {
+  private implementation?: E.SimpleApplication | undefined;
+  private readonly subTrees: Record<string, OverloadNode>;
+
+  public constructor() {
+    this.implementation = undefined;
+    this.subTrees = {};
+  }
+
+  /**
+   * Adds an overload to the tree structure considering this as the tree's root.
+   * @param argumentTypes a list of ArgumentTypes that would need to be provided in the same order to
+   * get the implementation.
+   * @param func the implementation for this overload.
+   */
+  public addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    this._addOverload([ ...argumentTypes ], func);
+  }
+
+  private _addOverload(argumentTypes: ArgumentType[], func: E.SimpleApplication): void {
+    const argument = argumentTypes.shift();
+    if (!argument) {
+      this.implementation = func;
+      return;
+    }
+    const str = argument;
+    if (!this.subTrees[str]) {
+      this.subTrees[str] = new OverloadNode();
+    }
+    this.subTrees[str]._addOverload(argumentTypes, func);
+  }
+
+  /**
+   * @param arg term to try and match to possible overloads of this node.
+   * @returns SearchStack a stack with top element the next node that should be asked for implementation or overload.
+   */
+  public getSubTreeWithArg(arg: E.TermExpression): SearchStack {
+    // The numbers a priority, a way to tell what type is more concrete and should be looked at first.
+    const matching: [number, OverloadNode][] = [];
+    for (const [ type, node ] of Object.entries(this.subTrees)) {
```

Comment:
For later (with extensionMap and decent suptyping and shit): we want to prioritize runtime performance over 'compile time' performance (read initialization). We could keep a sorted list of valid keys (in parallel with the subtrees), and then we only need to loop over that in order. No more sorting at runtime needed.

This would also be okay to do already, but a maybe bit overkill if you can do it with 3 if statements (up to you).

Reply:
I don't know whether we could work with an initialize phase. We'll have to see about that. For now I changed it to the if statements you asked for..